orders-accounting/src/cache.rs

```diff
@@ -0,0 +1,258 @@
+        let local_delta = self.data.ask_balances.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+
+    fn get_give_balance(&self, id: &OrderId) -> Result<Option<Amount>> {
+        let parent_supply = self.parent.get_give_balance(id).map_err(|_| Error::ViewFail)?;
+        let local_delta = self.data.give_balances.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+}
+
+impl<P: OrdersAccountingView> OrdersAccountingOperations for OrdersAccountingCache<P> {
+    fn create_order(&mut self, id: OrderId, data: OrderData) -> Result<OrdersAccountingUndo> {
+        if self.get_order_data(&id)?.is_some() {
+            return Err(Error::OrderAlreadyExists(id));
+        }
+
+        if self.get_ask_balance(&id)?.is_some() {
+            return Err(Error::OrderAlreadyExists(id));
+        }
+
+        let ask_value = data.ask().clone();
+        let give_value = data.give().clone();
+        let undo_data = self
+            .data
+            .order_data
+            .merge_delta_data_element(id, accounting::DataDelta::new(None, Some(data)))?;
+
+        self.data.ask_balances.add_unsigned(id, ask_value.amount())?;
+        self.data.give_balances.add_unsigned(id, give_value.amount())?;
+
+        Ok(OrdersAccountingUndo::CreateOrder(CreateOrderUndo {
+            id,
+            undo_data,
+            ask_balance: ask_value.amount(),
+            give_balance: give_value.amount(),
+        }))
+    }
+
+    fn withdraw_order(&mut self, id: OrderId) -> Result<OrdersAccountingUndo> {
+        let order_data = self
+            .get_order_data(&id)?
+            .ok_or(Error::AttemptedWithdrawNonexistingOrderData(id))?;
+        let ask_balance = self
+            .get_ask_balance(&id)?
+            .ok_or(Error::AttemptedWithdrawNonexistingAskBalance(id))?;
+        let give_balance = self
+            .get_give_balance(&id)?
+            .ok_or(Error::AttemptedWithdrawNonexistingGiveBalance(id))?;
+
+        let undo_data = self
+            .data
+            .order_data
+            .merge_delta_data_element(id, accounting::DataDelta::new(Some(order_data), None))?;
+
+        self.data.ask_balances.sub_unsigned(id, ask_balance)?;
+        self.data.give_balances.sub_unsigned(id, give_balance)?;
+
+        Ok(OrdersAccountingUndo::WithdrawOrder(WithdrawOrderUndo {
+            id,
+            undo_data,
+            ask_balance,
+            give_balance,
+        }))
+    }
+
+    fn fill_order(&mut self, id: OrderId, fill_value: OutputValue) -> Result<OrdersAccountingUndo> {
+        let filled_amount = calculate_fill_order(self, id, &fill_value)?;
+
+        let ask_balance = self.get_ask_balance(&id)?.ok_or(Error::OrderAskBalanceNotFound(id))?;
+        let give_balance =
+            self.get_give_balance(&id)?.ok_or(Error::OrderGiveBalanceNotFound(id))?;
+
```

Comment:
No check that the token vs coin pair are correct?

Reply:
checks are inside `calculate_fill_order` function

Reply:
Please ensure there's a test for this.

Reply:
added additional test